gpii/node_modules/lifecycleManager/src/UserLogonRequest.js

```diff
@@ -0,0 +1,211 @@
+    ],
+    listeners: {
+        "onCreate.logoutUser": {
+            listener: "{that}.handleRequest",
+            args: ["logout"]
+        }
+    }
+});
+
+/*
+ * The lifecycleManager logon request queue is used to hold the key-in and key-out requests.
+ * The entries in the action queue are of the format:
+ * {
+ *     gpiiKey: {String},
+ *     logonState: {String}  // enum: ["login", "logout"]
+ * }
+ * The action queue is run sequentially, and an item is considered "done" once the promise returned by the request
+ * is resolved.
+ */
+gpii.lifecycleManager.addToUserLogonRequestQueue = function (that, queue, item) {
+    var newItem = fluid.copy(item);
+    newItem.requestPromise = fluid.promise();
+    queue.push(newItem);
+    if (queue.length === 1) {
+        that.processUserLogonRequestQueue();
+    }
+    return newItem.requestPromise;
+};
+
+gpii.lifecycleManager.processUserLogonRequestQueue = function (that, queue) {
+    // Pick the first item and process. This item will be removed from the queue once the processing completes.
+    var item = queue[0];
+    var gradeName = gpii.lifecycleManager.userLogonRequest.handlerGrades[item.logonState];
+
+    if (!gradeName) {
+        fluid.fail("No handler grade found for user logon action - " + item.logonState);
+    };
+
+    that.events.onUserLogonRequestReceived.fire(gradeName, item.gpiiKey, item.requestPromise);
+
+    item.requestPromise.then(that.endProcessOneUserLogonRequest, that.endProcessOneUserLogonRequest);
+};
+
+gpii.lifecycleManager.endProcessOneUserLogonRequest = function (lifecycleManager) {
+    var queue = lifecycleManager.userLogonRequestQueue;
+
+    // The processing on the first item completes. Remove it from the queue.
+    queue.shift();
+
+    if (queue.length > 0) {
+        // process next item in the queue if it exists
+        lifecycleManager.processUserLogonRequestQueue(queue);
+    } else {
+        // if no actual user is logged in, login with "noUser"
+
+        // Use setTimeout() to work around the issue that kettle reports the error "Failing to resume callback for request xxx
+        // which has already concluded". The request that fails to resume callback is the user logon request that the current
+        // "endProcessOneUserLogonRequest" callback is attached with. It appears that starting the process of "noUser" keyin
+        // before the current callback completes will trigger this error.
+        setTimeout(function () {
+            var activeGpiiKeys = lifecycleManager.getActiveSessionGpiiKeys(); // find currently logged in user:
+
+            // Checking lifecycleManager.model.logonChange.inProgress is to make sure during the period to finish
+            // this request handling, there isn't another user logon request being added to the queue and in process.
+            if (activeGpiiKeys && activeGpiiKeys.length === 0 && !lifecycleManager.model.logonChange.inProgress) {
```

Comment:
Is this last check necessary? It seems that checking the queue length would be a more reliable test (for the time being)

Reply:
The last check is necessary due to the use of `setTimeout()` or `fluid.invoke()` that pulls out the function in it into an async process.

This check is to cover the case when the function within setTimeout runs simultaneously with the next logon request in the queue, in which case, the queue has become empty but the key in the next logon request hasn't been considered as active.

Reply:
But that isn't the case any more after the last refactoring, right? Since the queue always has at least one element in it so long as it is active at all?

Reply:
[This test](https://github.com/cindyli/universal/blob/GPII-3193/tests/shared/UserLogonRequestTestDefs.js#L243-L244) still can capture the case after the last refactoring. 

This example is, a keyin request adds to the queue right after "noUser" keys out. With "noUser" keyed out and before the keyin, the queue was empty and the function within setTimeout() was hit. When the keyin is added, the process of the keyin runs simultaneously with setTimeout() function, which will add "noUser" keyin back to the queue without the last check on `lifecycleManager.model.logonChange.inProgress`.

Reply:
It all seems quite risky. We should try to make sure that there is only one critical data structure being used at a time, and hopefully that structure is the queue. I don't feel completely confident that I understand the issue (at this point in the night) but I think that there is a fault in the implementation of performProximityTriggered, in that it explicitly waits for the conclusion of logoutPreviousUserPromise before it queues the login, whereas the whole point of the queued implementation was to ensure that we could register our login intention *immediately* so that the rest of the system can introspect on it. What I believe is that these two items should be added to the queue synchronously, and then hopefully races like this will go away. Please check over whether this seem sane to you ...

Reply:
How can you be such a genius, @amb26??!! Great capture! Adding keyout then keyin synchronously does solve the race condition. 

This pull request is ready for more reviews.

Thanks and cheers!

Reply:
Well, a real genius would have solved this problem by actually understanding it, whereas I merely reacted to what looked liked a collection of correlated smells : P But I'm glad that this smell led us to the problem since I'm sure I wouldn't have spotted the problem in performProximityTriggered by simply reading it during review.

Reply:
Sigh... This pull request is not ready yet. It was silly of me to forget about removing [the last check](https://github.com/cindyli/universal/blob/GPII-3193/gpii/node_modules/lifecycleManager/src/UserLogonRequest.js#L207). The race condition still persists without this check. I will re-think it through when my brain is clearer.

Reply:
@amb26, finally fixed the race condition. In the new commit, there are some adjustments to clean up the logic here and there, but the main trick is to fix [the scope of the function](https://github.com/cindyli/universal/blob/GPII-3193/gpii/node_modules/lifecycleManager/src/UserLogonRequest.js#L167-L190) in `fluid.invokeLater`. It's important to keep the queue change (`queue.shift()` call at line 171), and the actions resulted from the queue change  (line 173-189), in the same process.

This pull request is ready for another review.